packages/apollo-engine-reporting/src/plugin.ts

```diff
@@ -0,0 +1,403 @@
+          // source of bugs because integrations have different behavior here.
+          // On Node's HTTP module, request.url only includes the path
+          // (see https://nodejs.org/api/http.html#http_message_url)
+          // The same is true on Lambda (where we pass event.path)
+          // But on environments like Cloudflare we do get a complete URL.
+          host: null,
+          path: null,
+        });
+      }
+
+      let preflightDone: boolean = false;
+      function ensurePreflight() {
+        if (preflightDone) return;
+        preflightDone = true;
+
+        if (options.sendHeaders) {
+          if (requestContext.request.http && treeBuilder.trace.http) {
+            makeHTTPRequestHeaders(
+              treeBuilder.trace.http,
+              requestContext.request.http.headers,
+              options.sendHeaders,
+            );
+          }
+        }
+
+        if (metrics.persistedQueryHit) {
+          treeBuilder.trace.persistedQueryHit = true;
+        }
+        if (metrics.persistedQueryRegister) {
+          treeBuilder.trace.persistedQueryRegister = true;
+        }
+
+        // Generally, we'll get queryString here and not parsedQuery; we only get
+        // parsedQuery if you're using an OperationStore. In normal cases we'll
+        // get our documentAST in the execution callback after it is parsed.
+        queryString = requestContext.source;
+
+        if (requestContext.request.variables) {
+          treeBuilder.trace.details = makeTraceDetails(
+            requestContext.request.variables,
+            options.sendVariableValues,
+            queryString,
+          );
+        }
+
+        const clientInfo = generateClientInfo(requestContext);
+        if (clientInfo) {
+          // While clientAddress could be a part of the protobuf, we'll ignore it for
+          // now, since the backend does not group by it and Engine frontend will not
+          // support it in the short term
+          const { clientName, clientVersion, clientReferenceId } = clientInfo;
+          // the backend makes the choice of mapping clientName => clientReferenceId if
+          // no custom reference id is provided
+          treeBuilder.trace.clientVersion = clientVersion || '';
+          treeBuilder.trace.clientReferenceId = clientReferenceId || '';
+          treeBuilder.trace.clientName = clientName || '';
+        }
+      }
+
+      let endDone: boolean = false;
+      function didEnd() {
+        if (endDone) return;
+        endDone = true;
+        treeBuilder.stopTiming();
+
+        treeBuilder.trace.fullQueryCacheHit = !!metrics.responseCacheHit;
+        treeBuilder.trace.forbiddenOperation = !!metrics.forbiddenOperation;
+        treeBuilder.trace.registeredOperation = !!metrics.registeredOperation;
+
+        // If the user did not explicitly specify an operation name (which we
+        // would have saved in `executionDidStart`), but the request pipeline made
```

Comment:
You're not actually implementing this executionDidStart logic, which I think is specifically about detecting the operationName in the edge case where an operationName was explicitly mentioned by the client but that name isn't in the document. (If https://github.com/apollographql/apollo-server/pull/3465 ever merges that case will be reported more explicitly...) Too bad I never wrote a test for this case!

I might be wrong but I thiiiiiiiiiiiiink you can just do `requestContext.request.operationName || requestContext.operationName || ''`? Not sure why I didn't do that in the past instead of getting that value via execution args.

Reply:
Briefly: Precisely! This was introduced somewhat recently by #3894.

Providing a `logger` can be done at the `ApolloServer` level, `ApolloGateway` level and `EngineReportingOptions` level, _plus_ individual requests can assign over `logger` on the `requestContext` to create sub-scopes for a `logger` (if supported by the logger) to, e.g. assign `user_id` tags to log records.

Reply:
I'll need to fix this comment up, but if I recall, the existing pattern in the extension pipeline had to do some juggling because, maybe, the lack of `didResolveOperation`?  Do you think it would be reasonable to switch this to be captured within a `didResolveOperation` hook?  Invoked in the `requestPipeline` here:

https://github.com/apollographql/apollo-server/blob/6d4777da1f4dfaab3bf23523e9106e6c2abbab08/packages/apollo-server-core/src/requestPipeline.ts#L313-L335

Or are you suggesting that we want to be able to catch the error when a `requestContext.request.operationName` is specified.

Reply:
Right, didResolveOperation is great but there is a bit of a behavior change here for the case of "user specified an operation name that was not found".

Now I do think we should eventually finish up https://github.com/apollographql/apollo-server/pull/3465 which detects this case explicitly (as well as parse and validation failures) and groups all errors of that kind together, instead of acting like it's a legit operation. But maybe we can try to avoid the behavior change for now until that improvement is made?

I think all we need for that is to do the code I suggested above (or the other way round, `requestContext.operationName || requestContext.request.operationName || ''`) at some point on or after didResolveOperation.

Reply:
Addressed with your gracious help in b0948a826cc31bec79f0a54ea46fa2db884ac95b.  Thanks!